instance/models/mixins/storage.py

```diff
@@ -235,14 +253,23 @@ def provision_s3(self):
         """
         Create S3 Bucket if it doesn't exist
         """
-        if self.s3_access_key and self.s3_secret_access_key and self.s3_bucket_name:
-            self._create_bucket()
+        if not self.storage_type == self.S3_STORAGE:
+            return
+
+        if not self.s3_bucket_name:
+            self.s3_bucket_name = self.bucket_name
+
+        if not self.s3_access_key and not self.s3_secret_access_key:
+            self.create_iam_user()
+
+        self._create_bucket()
 
     def deprovision_s3(self):
         """
         Deprovision S3 by deleting S3 bucket and IAM user
         """
-        if not(self.s3_access_key or self.s3_secret_access_key or self.s3_bucket_name):
+        if not self.storage_type == self.S3_STORAGE or \
+                not(self.s3_access_key or self.s3_secret_access_key or self.s3_bucket_name):
```

Comment:
@viadanna maybe to avoid this long line, that is a bit hard to understand, we could add a private method:

def _is_s3_setup(self):
    return self.s3_access_key and self.s3_secret_access_key and self.s3_bucket_name

what do you think?

Reply:
That's an excellent idea!